File: python/572-Subtree-of-Another-Tree.py
from re import sub
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
   
        
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if not p or not q or p.val != q.val:
            return False
        return (self.isSameTree(p.left, q.left) and
                self.isSameTree(p.right, q.right))     
    
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root:
            return False
        
        if self.isSameTree(root, subRoot):
            return True
        if root.left and self.isSubtree(root.left, subRoot):
            return True
        if root.right and self.isSubtree(root.right, subRoot):
            return True
        return False
        
        
# Test
def main():
    test = Solution()
    n = 0
    print(f'{test.name()}')
    
# Run test script if this file is not being imported
if __name__ == '__main__':
    main()

File: javascript/572-Subtree-of-Another-Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
    let sameTree = (root, subRoot) => {
        if (!root && !subRoot) {
            return true;
        }
        if (!root || !subRoot || root.val != subRoot.val) {
            return false;
        }
        return sameTree(root.left, subRoot.left) && sameTree(root.right, subRoot.right)
    }

    let isSubtree = (root, subRoot) => {
        if (sameTree(root, subRoot)) {
            return true;
        }
        if (root && root.left && isSubtree(root.left, subRoot)) {
            return true;
        }
        if (root && root.right && isSubtree(root.right, subRoot)) {
            return true;
        }
        return false;
    };